dll/win32/shell32/folders/CFindFolder.cpp

```diff
@@ -0,0 +1,609 @@
+            continue;
+
+        PathCombineW(szPath, lpPath, FindData.cFileName);
+
+        if (FindData.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+        {
+            CStringW status;
+            status.Format(IDS_SEARCH_FOLDER, FindData.cFileName);
+            PostMessageW(pSearchData->hwnd, WM_SEARCH_UPDATE_STATUS, 0, (LPARAM) StrDupW(status.GetBuffer()));
+
+            uTotalFound += RecursiveFind(szPath, pSearchData);
+        }
+        else if ((pSearchData->szFileName.IsEmpty() || PathMatchSpecW(FindData.cFileName, pSearchData->szFileName))
+                && (pSearchData->szQueryA.IsEmpty() || SearchFile(szPath, pSearchData)))
+        {
+            uTotalFound++;
+            PostMessageW(pSearchData->hwnd, WM_SEARCH_ADD_RESULT, 0, (LPARAM) StrDupW(szPath));
+        }
+    }
+
+    if (hFindFile != INVALID_HANDLE_VALUE)
+        FindClose(hFindFile);
+
+    return uTotalFound;
+}
+
+DWORD WINAPI CFindFolder::SearchThreadProc(LPVOID lpParameter)
+{
+    _SearchData *data = static_cast<_SearchData*>(lpParameter);
+
+    data->pFindFolder->NotifyConnections(DISPID_SEARCHSTART);
+
+    UINT uTotalFound = RecursiveFind(data->szPath, data);
+
+    data->pFindFolder->NotifyConnections(DISPID_SEARCHCOMPLETE);
+
+    CStringW status;
+    status.Format(IDS_SEARCH_FILES_FOUND, uTotalFound);
+    ::PostMessageW(data->hwnd, WM_SEARCH_UPDATE_STATUS, 0, (LPARAM) StrDupW(status.GetBuffer()));
+
+    delete data;
+
+    return 0;
+}
+
+void CFindFolder::NotifyConnections(DISPID id)
+{
+    DISPPARAMS dispatchParams = {0};
+    CComDynamicUnkArray &subscribers =
+        IConnectionPointImpl<CFindFolder, &DIID_DSearchCommandEvents>::m_vec;
+    for (IUnknown** pSubscriber = subscribers.begin(); pSubscriber < subscribers.end(); pSubscriber++)
+    {
+        if (!*pSubscriber)
+            continue;
+
+        CComPtr<IDispatch> pDispatch;
+        HRESULT hResult = (*pSubscriber)->QueryInterface(IID_PPV_ARG(IDispatch, &pDispatch));
+        if (!FAILED_UNEXPECTEDLY(hResult))
+            pDispatch->Invoke(id, GUID_NULL, 0, DISPATCH_METHOD, &dispatchParams, NULL, NULL, NULL);
+    }
+}
+
+LRESULT CFindFolder::StartSearch(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    if (!lParam)
+        return 0;
+
+    // Clear all previous search results
+    UINT uItemIndex;
+    m_shellFolderView->RemoveObject(NULL, &uItemIndex);
+
+    _SearchData* pSearchData = new _SearchData();
+    pSearchData->pFindFolder = this;
+    pSearchData->hwnd = m_hWnd;
+
+    SearchStart *pSearchParams = (SearchStart *) lParam;
+    pSearchData->szPath = pSearchParams->szPath;
+    pSearchData->szFileName = pSearchParams->szFileName;
+    pSearchData->szQueryA = pSearchParams->szQuery;
+    pSearchData->szQueryW = pSearchParams->szQuery;
+    SHFree(pSearchParams);
+
+    if (m_hStopEvent)
+        SetEvent(m_hStopEvent);
+    pSearchData->hStopEvent = m_hStopEvent = CreateEvent(NULL, TRUE, FALSE, NULL);
```

Comment:
Leaking `m_hStopEvent`.
But since both the search thread and the `main` thread are using the same handle,
it can only be deleted when the search thread is done. (So probably the search thread should delete the object when it is done, and the main thread should clear the `m_hStopEvent` when the search is done)